### PR TITLE
36 download database files

### PR DESCRIPTION
**Pull request by nilsoberg**

### CODE REVIEW

File: scripts/download_file.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    String comparison of local and remote MD5 checksums
+    """
+    if remote_md5 != local_md5:
+        print(f"Remote MD5 {remote_md5} doesn't match local MD5 {local_md5}")
+        return False
+    else:
+        return True
+
+
+
```
Comment (1ndy) [inline scripts/download_file.py:119]: `if __name__ == '__main__':`

File: scripts/download_file.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
```
Comment (1ndy) [inline scripts/download_file.py:1]: Consider adding a few lines that explain what this script is for

File: scripts/download_file.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import argparse
+import urllib.request
+import urllib.error
+import os
+import hashlib
+import sys
+
+
```
Comment (1ndy) [inline scripts/download_file.py:9]: `def parse_args(): argparse.Namespace` so that this does not run every time the file executes

File: scripts/download_file.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        response = urllib.request.urlopen(remote_file)
+    except urllib.error.HTTPError as e:
+        print(f"Unable to urlopen {remote_file}: {e}")
+        return False
+    file_contents = response.read()
+    file_contents = str(file_contents.decode('utf-8'))
+    return file_contents
+
+
+def get_file_list(args):
```
Comment (1ndy) [inline scripts/download_file.py:40]: be explicit about which args you are using: only `remote_dir` and `remote_file`

File: scripts/download_file.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    os.mkdir(args.local_dir)
+
+
+
+
+
+
+def get_remote_file_contents(remote_file):
+    """
+    Downloads the file as a string and return to user; if HTTP error then returns False
```
Comment (1ndy) [inline scripts/download_file.py:28]: add parameter and return sections to every docstring and consider type-annotating your functions

File: scripts/download_file.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    """
+    try:
+        response = urllib.request.urlopen(remote_file)
+    except urllib.error.HTTPError as e:
+        print(f"Unable to urlopen {remote_file}: {e}")
+        return False
```
Comment (1ndy) [inline scripts/download_file.py:34]: I think it would be a little more idiomatic to return `None` and then test for `is not None` wherever this function is used

File: scripts/download_file.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    lines = file_contents.splitlines()
+    checksums = {}
+    files = []
+    for line in lines:
+        parts = line.split(" *")
+        files.append(parts[1])
+        checksums[parts[1]] = parts[0]
+
+    return files, checksums
+
```
Comment (1ndy) [inline scripts/download_file.py:59]: `splitlines` will return a copy of the data. It would be more memory efficient to iterate over the decoded buffer directly. See if you can make some changes in `get_remote_file_contents` so that you get back a [`StringIO`](https://docs.python.org/3/library/io.html#text-i-o) object. Also, it looks like you maintain the list of filenames both in an array and as the keys of the dict. Could you just use the dict and call `.keys()` when the filenames are needed?

File: scripts/download_file.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+def get_remote_checksum(args):
+    """
+    Return the MD5 hash from the MD5 file corresponding to the remote file
+    """
+    remote_file = args.remote_dir + "/" + args.remote_file + ".md5"
+
+    file_contents = get_remote_file_contents(remote_file)
+    if not file_contents:
+        return False
```
Comment (1ndy) [inline scripts/download_file.py:69]: Return `None`

File: scripts/download_file.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+
+file_list, checksums = get_file_list(args)
+
+if len(file_list) > 0:
+    temp_files = []
+    for fname in file_list:
+        temp_local_file = os.path.join(args.local_dir, fname)
+        temp_files.append(temp_local_file)
+        remote_url = args.remote_dir + "/" + fname
```
Comment (1ndy) [inline scripts/download_file.py:128]: Consider using `urllib.parse.urljoin`

File: scripts/download_file.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    file_contents = response.read()
+    file_contents = str(file_contents.decode('utf-8'))
+    return file_contents
+
+
+def get_file_list(args):
+    """
+    Retrieves a list of files on the remote that need to be fetched
+    """
+    remote_file = args.remote_dir + "/" + args.remote_file + ".file_list"
```
Comment (1ndy) [inline scripts/download_file.py:44]: Consider using `urllib.parse.urljoin`

File: scripts/download_file.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    Download and save a file specified by url to the destination file
+    """
+    try:
+        u = urllib.request.urlopen(url)
+    except urllib.error.HTTPError as e:
+        print(f"Unable to urlopen {url}: {e}")
+        return False
+
+    with open(dest_file, "wb") as f:
+        block_sz = 8192
```
Comment (1ndy) [inline scripts/download_file.py:89]: Consider making this a "global const" at the top of the file

```
BLOCK_SIZE = 8192
```

File: scripts/download_file.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        for fname in temp_files:
+            with open(fname, "rb") as fh:
+                for chunk in iter(lambda: fh.read(4096), b""):
+                    afh.write(chunk)
+
+    #remote_md5 = get_remote_checksum(args)
+    #local_md5 = calculate_md5(args.local_file)
+
+else:
+    remote_url = args.remote_dir + "/" + args.remote_file
```
Comment (1ndy) [inline scripts/download_file.py:148]: Consider using urllib.parse.urljoin

File: scripts/download_file.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+        download_file(remote_url, temp_local_file)
+        local_md5 = calculate_md5(temp_local_file)
+        if not compare_md5(checksums[fname], local_md5):
+            print(f"File {fname} was not downloaded correctly; checksums do not match")
+            sys.exit(1)
+
+    with open(args.local_file, "wb") as afh:
+        for fname in temp_files:
+            with open(fname, "rb") as fh:
+                for chunk in iter(lambda: fh.read(4096), b""):
```
Comment (1ndy) [inline scripts/download_file.py:141]: You could also do
```
while chunk := fh.read(4096):
```
I don't have a strong opinion. But do consider assigning 4096 to a global const style variable

File: scripts/download_file.py (hunk 9 above)
Comment (nilsoberg) [reply scripts/download_file.py:128]: `urllib.parse.urljoin` doesn't do what I would expect it to do based on it's name, so rather than appending `fname` to `args.remote_dir`, it replaces the last element in the `args.remote_dir` path with `fname`.

File: scripts/download_file.py (hunk 7 above)
Comment (nilsoberg) [reply scripts/download_file.py:100]: The data returned by `get_remote_file_contents` will be quite small.  It is simply the output of running `md5sum dir/files* > dir/dir.file_list`.  However, I think what you suggest is the correct approach.  I will investigate.  Regarding the array/dict, the files need to be concatenated in order.  I believe that dictionaries in Python are ordered so yes, I can return one dict rather than an array and dict.

Comment (nilsoberg) [reply scripts/download_file.py]: @1ndy The default Python install on some OS is Python 3.6 and implicitly ordered dictionaries were introduced after 3.6.  Do you think its safe to require a specific version of Python in order for the script to run?  (By using `sys.version_info` for example.)  It looks like KBase itself is based on a Python 3.8 implementation.

File: scripts/download_file.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+            path to a file on the local file system
+
+    Returns
+    -------
+        MD5 hash of the file
+    """
+
+    hash_md5 = hashlib.md5()
+    with open(file_path, "rb") as fh:
+        while chunk := fh.read(BLOCK_SIZE):
```
Comment (nilsoberg) [inline scripts/download_file.py:167]: The `:=` operator is not supported in Python 3.6.  On an architectural level should we support older versions of Python 3?

File: scripts/download_file.py (hunk 7 above)
Comment (1ndy) [reply scripts/download_file.py]: Ah right, yeah I wouldn't try to force a specific python version just for this feature, you can use [`OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict) instead

File: scripts/download_file.py (hunk 14 above)
Comment (1ndy) [reply scripts/download_file.py:167]: You kind of need to support older python within this script because it is used inside KBase. If that were not true, I would say that only supporting active versions of python (which [as of now is 3.9+](https://devguide.python.org/versions/#python-release-cycle)) would be the way to go. I think `:=` is a python3.8 feature but I might be wrong.

I think the code you previously had is fine given the version constraint.

File: scripts/download_file.py (hunk 10 above)
Comment (nilsoberg) [reply scripts/download_file.py]: I looked into pathlib; it's for file system paths not URLs.  I elected to keep using string concatenation here because the application is very specific.  URL separators are forward slashes (`/`) so I think doing the concatenation here is acceptable.

**Review by rbdavid — APPROVED** [review]
Minor comments. Nothing that needs to be patched. Looks good

File: scripts/download_file.py (hunk 10 above)
Comment (rbdavid) [reply scripts/download_file.py]: Linux paths use forward slashes, so pathlib works as expected in my working environment. I am happily naive about how Windows formats paths.

File: scripts/download_file.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+if __name__ == '__main__':
+
+    args = parse_args()
+
+    # Check if the remote file is actually a collection of files that were split up
+    file_list = get_file_list(args.remote_dir, args.remote_file)
+
+    # If the remote was split, then download all of the split files
+    if file_list is not None:
```
[the remark this answers is not in the document]
Comment (nilsoberg) [reply scripts/download_file.py:205]: Thanks - will keep that in mind.  Staying with `is not None` in order to keep meaning of return values consistent (e.g. possible values are dict size > 0, dict size == 0, HTTP error).

File: scripts/download_file.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+        if not os.path.isdir(args.local_dir):
+            os.mkdir(args.local_dir)
```
[the remark this answers is not in the document]
Comment (nilsoberg) [reply scripts/download_file.py]: Changed

**Review by rbdavid — APPROVED** [review]
Looks good to me.

**Review by 1ndy — APPROVED** [review]
